src/rdkafka_queue.c

```diff
@@ -516,8 +516,32 @@ int rd_kafka_q_serve (rd_kafka_q_t *rkq, int timeout_ms,
 	return cnt;
 }
 
+/**
+ * @brief Filter out and destroy outdated messages.
+ *
+ * @returns Returns the number of valid messages.
+ *
+ * @locality Any thread.
+ */
+static size_t rd_kafka_purge_outdated_messages (int32_t version,
+                                  rd_kafka_message_t **rkmessages,
+                                  int cnt) {
+        rd_kafka_op_t *rko;
+        size_t valid_count = 0;
+        int i;
+
+        for (i = 0; i < cnt; i++) {
+                rko = rkmessages[i]->_private;
+                if (rd_kafka_op_version_outdated(rko, version)) {
+                        /* This also destroys the corresponding rkmessage. */
+                        rd_kafka_op_destroy(rko);
+                } else {
```

Comment:
Perhaps as an optimization to avoid a write shuffle:  `else if (i > valid_count++)`

Reply:
Hi Magnus, thanks for pointing this out! I think copy should before valid_count + 1, so I change this part to the below logic:

if (rd_kafka_op_version_outdated(rko, version)) {
                        /* This also destroys the corresponding rkmessage. */
                        rd_kafka_op_destroy(rko);
} else if ((size_t)i > valid_count) {
                        rkmessages[valid_count++] = rkmessages[i];
} else {
                        valid_count++;
}